SimExLite/PMICalculators/SimpleScatteringPMICalculator.py

```diff
@@ -0,0 +1,11147 @@
+    if wavefront_data.mapping_type == WPGFormat:
+        wavefront_fn = wavefront_data.filename
+    else:
+        temp_name = next(tempfile._get_candidate_names())
+        wavefront_fn = "input_wavefront_" + temp_name + ".h5"
+        wavefront_data.write(wavefront_fn, WPGFormat)
+    return wavefront_fn
+
+
+def SampleData_to_atoms_dict(sample_data: SampleData) -> dict:
+    """Read the sampleData object into an atoms_dict for the backengine"""
+    atoms_dict = {
+        "Z": [],  # Atomic number.
+        "r": [],  # Cartesian coordinates.
+        "selZ": {},  # Abundance of each element.
+        "N": 0,  # Number of atoms.
+    }
+    input_dict = sample_data.get_data()
+    atoms_dict["Z"] = input_dict["atomic_numbers"]
+    atoms_dict["r"] = input_dict["positions"] * 1e-10
+    atoms_dict["N"] = len(input_dict["atomic_numbers"])
+    for sel_Z in numpy.unique(atoms_dict["Z"]):
+        atoms_dict["selZ"][sel_Z] = numpy.nonzero(atoms_dict["Z"] == sel_Z)[0]
+    return atoms_dict
+
+
+class PMIScattering(object):
+    def __init__(self):
+        self.g_s2e = {}
+        self.g_dbase = {}
+
+        self.f_s2e_setup()
+
+    def f_s2e_setup(self):
+        self.g_s2e["setup"] = dict()
+        self.g_s2e["sys"] = dict()
+        self.g_s2e["setup"]["num_digits"] = 7
+        self.g_s2e["steps"] = 100
+        self.g_s2e["maxZ"] = 100
+
+    ##############################################################################
+
+    def f_save_info(self):
+        pmi_file = self.g_s2e["setup"]["pmi_out"]
+
+        grp = "/info"
+        ###    print pmi_file , grp
+        xfp = h5py.File(pmi_file, "a")
+        try:
+            grp_hist_parent = xfp.create_group(grp)
+        except:
+            1
+
+        #    xfp[ grp + '/Sq_free' ] = g_dbase['Sq_free']
+        xfp.close()
+
+    ##############################################################################
+
+    def f_init_random(self):
+        random.seed(self.g_s2e["id"])
+
+    ##############################################################################
+
+    def f_dbase_Zq2id(self, a_Z, a_q):
+        return (a_Z * (a_Z + 1)) // 2 - 1 * 1 + a_q
+
+    ##############################################################################
+
+    def f_dbase_setup(self):
+        # print '   Update database!!!'
+        xdbase = load_ff_database()
+
+        g_dbase = dict()
+
+        #   q               ->   sin(theta/2)/lambda
+        #   au, exp(i*q*r)  ->   1/Angstrom, exp( 2*pi*q*r)
+        g_dbase["halfQ"] = xdbase[:, 0] = xdbase[:, 0] / (
+            2.0 * numpy.pi * 0.529177206 * 2.0
+        )
+
+        maxZ = self.g_s2e["maxZ"]  # 99
+        numQ = len(g_dbase["halfQ"])
+        g_dbase["ff"] = numpy.zeros((self.f_dbase_Zq2id(maxZ, maxZ) + 1, numQ))
+        ii = 0
+        for ZZ in range(1, maxZ + 1):
+            for qq in range(ZZ + 1):
+                g_dbase["ff"][ii, :] = xdbase[:, ZZ] * (ZZ - qq) / (ZZ * 1.0)
+                ii = ii + 1
```

Comment:
`ii += 1` is a bit faster and more elegant (not like it matters much here).

Reply:
Changed as suggested, thanks. Is there any benchmark showing how faster it could be?

Reply:
Just checked and the difference actually seems to be minimal:
![image](https://user-images.githubusercontent.com/10786406/162994810-0f45a14c-5b32-4192-8b40-ac936a65c871.png)